{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2a165196",
   "metadata": {},
   "source": [
    "Check test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ec87d86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>...</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>canonical_center</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "      <th>prot_ref_seq</th>\n",
       "      <th>prot_alt_seq</th>\n",
       "      <th>label</th>\n",
       "      <th>CHROM_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>716893</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_153269.3(C20orf96):c.183A&gt;G (p.Gln61=)</td>\n",
       "      <td>140680</td>\n",
       "      <td>C20orf96</td>\n",
       "      <td>HGNC:16227</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>Q</td>\n",
       "      <td>caA/caG</td>\n",
       "      <td>289563</td>\n",
       "      <td>CTGGATGCAAAAGTATCCAGCAGAAGACTTCAAGACCTTACAAAAT...</td>\n",
       "      <td>CTGGATGCAAAAGTATCCAGCAGAAGACTTCAAGACCTTACAAAAT...</td>\n",
       "      <td>SGTHSIVQEFQVPDYVPWQQSKQETKPSTLPPVQQANSLHTSKMKT...</td>\n",
       "      <td>SGTHSIVQEFQVPDYVPWQQSKQETKPSTLPPVQQANSLHTSKMKT...</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>742344</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006943.4(SOX12):c.636C&gt;G (p.Ala212=)</td>\n",
       "      <td>6666</td>\n",
       "      <td>SOX12</td>\n",
       "      <td>HGNC:11198</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>...</td>\n",
       "      <td>212</td>\n",
       "      <td>A</td>\n",
       "      <td>gcC/gcG</td>\n",
       "      <td>326560</td>\n",
       "      <td>CAAAAAGAGCAAGGGGGCGCCCGCCAAGGCGCGGCCCCGCCCCCCC...</td>\n",
       "      <td>CAAAAAGAGCAAGGGGGCGCCCGCCAAGGCGCGGCCCCGCCCCCCC...</td>\n",
       "      <td>PEDDDEDDDEELLEVRLVETPGRELWRMVPAGRAARGQAERAQGPS...</td>\n",
       "      <td>PEDDDEDDDEELLEVRLVETPGRELWRMVPAGRAARGQAERAQGPS...</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>742345</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006943.4(SOX12):c.728A&gt;G (p.Glu243Gly)</td>\n",
       "      <td>6666</td>\n",
       "      <td>SOX12</td>\n",
       "      <td>HGNC:11198</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>...</td>\n",
       "      <td>243</td>\n",
       "      <td>E/G</td>\n",
       "      <td>gAg/gGg</td>\n",
       "      <td>326652</td>\n",
       "      <td>CTGGCCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGG...</td>\n",
       "      <td>CTGGCCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGG...</td>\n",
       "      <td>GRAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAE...</td>\n",
       "      <td>GRAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAE...</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>705433</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006943.4(SOX12):c.732G&gt;A (p.Thr244=)</td>\n",
       "      <td>6666</td>\n",
       "      <td>SOX12</td>\n",
       "      <td>HGNC:11198</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>T</td>\n",
       "      <td>acG/acA</td>\n",
       "      <td>326656</td>\n",
       "      <td>CCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGGGGCG...</td>\n",
       "      <td>CCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGGGGCG...</td>\n",
       "      <td>RAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAEE...</td>\n",
       "      <td>RAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAEE...</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>716917</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001323682.2(NRSN2):c.30C&gt;T (p.Ser10=)</td>\n",
       "      <td>80023</td>\n",
       "      <td>NRSN2</td>\n",
       "      <td>HGNC:16229</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>S</td>\n",
       "      <td>agC/agT</td>\n",
       "      <td>349673</td>\n",
       "      <td>GGCCTCCAGAGTAAACATGCACATTTATTGACACCTCCAATTGCCA...</td>\n",
       "      <td>GGCCTCCAGAGTAAACATGCACATTTATTGACACCTCCAATTGCCA...</td>\n",
       "      <td>XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXMMPSC...</td>\n",
       "      <td>XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXMMPSC...</td>\n",
       "      <td>0</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AlleleID                       Type  \\\n",
       "0    716893  single nucleotide variant   \n",
       "1    742344  single nucleotide variant   \n",
       "2    742345  single nucleotide variant   \n",
       "3    705433  single nucleotide variant   \n",
       "4    716917  single nucleotide variant   \n",
       "\n",
       "                                        Name  GeneID GeneSymbol     HGNC_ID  \\\n",
       "0  NM_153269.3(C20orf96):c.183A>G (p.Gln61=)  140680   C20orf96  HGNC:16227   \n",
       "1    NM_006943.4(SOX12):c.636C>G (p.Ala212=)    6666      SOX12  HGNC:11198   \n",
       "2  NM_006943.4(SOX12):c.728A>G (p.Glu243Gly)    6666      SOX12  HGNC:11198   \n",
       "3    NM_006943.4(SOX12):c.732G>A (p.Thr244=)    6666      SOX12  HGNC:11198   \n",
       "4   NM_001323682.2(NRSN2):c.30C>T (p.Ser10=)   80023      NRSN2  HGNC:16229   \n",
       "\n",
       "  ClinicalSignificance  ClinSigSimple     PhenotypeIDS PhenotypeList  ...  \\\n",
       "0               Benign              0  MedGen:C3661900  not provided  ...   \n",
       "1               Benign              0  MedGen:C3661900  not provided  ...   \n",
       "2               Benign              0  MedGen:C3661900  not provided  ...   \n",
       "3               Benign              0  MedGen:C3661900  not provided  ...   \n",
       "4               Benign              0  MedGen:C3661900  not provided  ...   \n",
       "\n",
       "  Protein_position Amino_acids   Codons canonical_center  \\\n",
       "0               61           Q  caA/caG           289563   \n",
       "1              212           A  gcC/gcG           326560   \n",
       "2              243         E/G  gAg/gGg           326652   \n",
       "3              244           T  acG/acA           326656   \n",
       "4               10           S  agC/agT           349673   \n",
       "\n",
       "                                             ref_seq  \\\n",
       "0  CTGGATGCAAAAGTATCCAGCAGAAGACTTCAAGACCTTACAAAAT...   \n",
       "1  CAAAAAGAGCAAGGGGGCGCCCGCCAAGGCGCGGCCCCGCCCCCCC...   \n",
       "2  CTGGCCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGG...   \n",
       "3  CCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGGGGCG...   \n",
       "4  GGCCTCCAGAGTAAACATGCACATTTATTGACACCTCCAATTGCCA...   \n",
       "\n",
       "                                             alt_seq  \\\n",
       "0  CTGGATGCAAAAGTATCCAGCAGAAGACTTCAAGACCTTACAAAAT...   \n",
       "1  CAAAAAGAGCAAGGGGGCGCCCGCCAAGGCGCGGCCCCGCCCCCCC...   \n",
       "2  CTGGCCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGG...   \n",
       "3  CCGCGGGGGCCGCCGAGCAGCGGGAGGGCCTTTGGGGGGCGGGGCG...   \n",
       "4  GGCCTCCAGAGTAAACATGCACATTTATTGACACCTCCAATTGCCA...   \n",
       "\n",
       "                                        prot_ref_seq  \\\n",
       "0  SGTHSIVQEFQVPDYVPWQQSKQETKPSTLPPVQQANSLHTSKMKT...   \n",
       "1  PEDDDEDDDEELLEVRLVETPGRELWRMVPAGRAARGQAERAQGPS...   \n",
       "2  GRAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAE...   \n",
       "3  RAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAEE...   \n",
       "4  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXMMPSC...   \n",
       "\n",
       "                                        prot_alt_seq  label  CHROM_norm  \n",
       "0  SGTHSIVQEFQVPDYVPWQQSKQETKPSTLPPVQQANSLHTSKMKT...      0       chr20  \n",
       "1  PEDDDEDDDEELLEVRLVETPGRELWRMVPAGRAARGQAERAQGPS...      0       chr20  \n",
       "2  GRAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAE...      0       chr20  \n",
       "3  RAARGQAERAQGPSGEGAAAAAAASPTPSEDEEPEEEEEEAAAAEE...      0       chr20  \n",
       "4  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXMMPSC...      0       chr20  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "parquet_path = r\"test.parquet\"\n",
    "df = pd.read_parquet(parquet_path, engine=\"pyarrow\")\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b1607d09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID',\n",
       "       'ClinicalSignificance', 'ClinSigSimple', 'PhenotypeIDS',\n",
       "       'PhenotypeList', 'Origin', 'OriginSimple', 'Assembly',\n",
       "       'ChromosomeAccession', 'CHROM', 'Start', 'Stop', 'ReviewStatus',\n",
       "       'VariationID', 'POS', 'REF', 'ALT', 'Uploaded_variation', 'Consequence',\n",
       "       'HGVSc', 'Feature', 'Protein_id', 'Protein_position', 'Amino_acids',\n",
       "       'Codons', 'canonical_center', 'ref_seq', 'alt_seq', 'prot_ref_seq',\n",
       "       'prot_alt_seq', 'label', 'CHROM_norm'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78fa3f7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ MAGPIE minimal input saved to: magpie_input.csv\n",
      "  Chr   Start     End Ref Alt\n",
      "0  20  289563  289563   T   C\n",
      "1  20  326560  326560   C   G\n",
      "2  20  326652  326652   A   G\n",
      "3  20  326656  326656   G   A\n",
      "4  20  349673  349673   C   T\n",
      "Shape: (4562, 5)\n"
     ]
    }
   ],
   "source": [
    "# Chr\n",
    "if \"CHROM_norm\" in df.columns:\n",
    "    df[\"Chr\"] = (\n",
    "        df[\"CHROM_norm\"]\n",
    "        .astype(str)\n",
    "        .str.replace(\"chr\", \"\", regex=False)\n",
    "    )\n",
    "elif \"CHROM\" in df.columns:\n",
    "    df[\"Chr\"] = (\n",
    "        df[\"CHROM\"]\n",
    "        .astype(str)\n",
    "        .str.replace(\"chr\", \"\", regex=False)\n",
    "    )\n",
    "else:\n",
    "    raise ValueError(\"❌ Không tìm thấy CHROM hoặc CHROM_norm\")\n",
    "\n",
    "# Position\n",
    "if \"POS\" not in df.columns:\n",
    "    raise ValueError(\"❌ Thiếu cột POS\")\n",
    "\n",
    "df[\"Start\"] = df[\"POS\"]\n",
    "df[\"End\"] = df[\"POS\"]  # SNV → Start = End\n",
    "\n",
    "# REF - ALT\n",
    "if \"REF\" not in df.columns or \"ALT\" not in df.columns:\n",
    "    raise ValueError(\"❌ Thiếu REF hoặc ALT\")\n",
    "\n",
    "df[\"Ref\"] = df[\"REF\"]\n",
    "df[\"Alt\"] = df[\"ALT\"]\n",
    "\n",
    "\n",
    "df_magpie = df[[\"Chr\", \"Start\", \"End\", \"Ref\", \"Alt\"]].copy()\n",
    "\n",
    "# Save\n",
    "output_path = \"magpie_input.csv\"\n",
    "df_magpie.to_csv(output_path, index=False)\n",
    "\n",
    "print(\"MAGPIE minimal input saved to:\", output_path)\n",
    "print(df_magpie.head())\n",
    "print(\"Shape:\", df_magpie.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd02e9fa",
   "metadata": {},
   "source": [
    "Clone git"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cebc52c1",
   "metadata": {},
   "source": [
    "git clone https://github.com/ShenLab-Genomics/MAGPIE\n",
    "\n",
    "cd MAGPIE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78afb3b9",
   "metadata": {},
   "source": [
    "CREATE CONDA ENVIRONMENT (RECOMMENDED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e497c28",
   "metadata": {},
   "source": [
    "Linux: \n",
    "\n",
    "    conda env create -f magpie.yml\n",
    "\n",
    "    conda activate magpie\n",
    "\n",
    "Windows:\n",
    "\n",
    "    conda create -n magpie python=3.9 -y\n",
    "\n",
    "    conda activate magpie\n",
    "\n",
    "    pip install -r requirements.txt\n",
    "\n",
    "    conda install -c conda-forge pyvcf matplotlib seaborn plotnine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06faae9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install openfe==0.0.7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9354de",
   "metadata": {},
   "source": [
    "source magpie.sh --mode pred --test_file D:\\my_project\\Bio_paper\\test\\magpie_input_annotated.csv --file_state annotated --visualization\n",
    "\n",
    "source magpie.sh --mode pred \\\n",
    "  --test_file /d/my_project/Bio_paper/test/magpie_input.csv \\\n",
    "  --file_state unannotated \\\n",
    "  --visualization\n",
    "\n",
    "bash -x magpie.sh --mode pred \\\n",
    "  --test_file /d/my_project/Bio_paper/test/magpie_input.csv\n",
    "  --file_state unannotated \\\n",
    "  --visualization"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ai_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
